### ISSUE TITLE
Bug: chainrules autograd-next rejects untracked constants that carry a foreign context

### ISSUE THREAD

**Issue by shinaoka** [tensor4all/tenferro-rs#396]
## Summary

`extern/chainrules`'s autograd-next API rejects valid multi-input operations when an operand is logically a constant (`requires_grad() == false`) but still carries a different `AutogradContext`.

This happens for both:

- `Variable::new_in(value, foreign_ctx)` with default `requires_grad = false`
- values that were once tracked and then toggled with `requires_grad_(false)`

At minimum this affects `autograd::add`, and the same context-merge helper also gates `autograd::record_op`.

## Repro

Verified with an external consumer crate:

```bash
cargo run --offline --manifest-path target/tmp/bug-chainrules-disabled-constant/Cargo.toml
```

Observed output:

```text
baseline_grad=Some(4.0)
constant_requires_grad=false constant_context_id=Some(2) constant_node_id=Some(NodeId(0))
tracked_plus_plain_constant=ok requires_grad=true context_id=Some(1) node_id=Some(NodeId(2)) value=5
tracked_plus_disabled=err invalid autodiff argument: mixed autograd contexts in one operation; use Variable::new_in(..., same_ctx)
fresh_constant_requires_grad=false fresh_constant_context_id=Some(2) fresh_constant_node_id=None
tracked_plus_fresh_constant=err invalid autodiff argument: mixed autograd contexts in one operation; use Variable::new_in(..., same_ctx)
```

## Expected behavior

If an operand has `requires_grad() == false`, it should behave as a constant and should not participate in cross-context conflict detection.

These should succeed:

- `autograd::add(&tracked, &Variable::new_in(constant, foreign_ctx))`
- `autograd::add(&tracked, &formerly_tracked.requires_grad_(false)?)`

The output should stay on the tracked operand's context, just as it already does for a plain context-free constant.

## Actual behavior

Multi-input ops reject those constants as if they were tracked foreign-graph inputs, even though they do not require gradients.

## Suspected root cause

Three pieces combine here:

1. `Variable::new_in(...)` stores a context even when `requires_grad` is false.
2. `Variable::requires_grad_` flips the boolean but does not clear `context` or `node_id` when disabling tracking.
3. `merge_context_for_multi_op` iterates over every attached context, not only over tracked inputs.

That makes an untracked constant look like a mixed-context operand and causes an `InvalidArgument` before the operation can be recorded.

## Source references

- `extern/chainrules/src/lib.rs:1396`
- `extern/chainrules/src/lib.rs:1470`
- `extern/chainrules/src/lib.rs:1800`
- `extern/chainrules/src/lib.rs:1872`
- `extern/chainrules/src/lib.rs:1911`

## Report

- `docs/test-reports/agentic-bug-sweep/bug-sweep-20260308-180733.md`
